# Laboratorio Optimización No. 8

Daniel Alvarez - 201911320  
Maria Jose Vieda - 201923359

### Importacion de Librerias:

In [266]:
import numpy as np
import time
import pandas as pd
from IPython.display import display
from scipy.optimize import linprog, minimize
import time
from scipy.linalg import solve
from collections import Counter

# Punto 1
G&L es una pequeña fabrica de equipos y accesorios de golf. Donde el dueño de dicha fabrica busca poder maximinar la distribución total para cada utilidad. Siendo estos, dos productos blasa deluxe y bolsa estandar con costos de utilidad de $9 y $7 respectivamente. Durante los siguientes tres meses estarán disponibles 630 horas de tiempo de corte y teñido, 600 de tiempo de costura, 708 horas de tiempo de terminado y 135 horas de tiempo de inspección y empaque para la producción de las bolsas de golf.

### Variables de desición:  
> Para tener la utilidad general de la suma de las bolsas tenemos la siguiente función de costo:  
$$\max 7x_1 + 9x_2 = P(x_1,x_2)$$
$$x_{1} \rightarrow Estandar$$
$$x_{2} \rightarrow Deluxe$$
$$\min -7x_1 - 9x_2 = P(x_1,x_2)$$
> Donde se tiene en general restricciones relacionadas con el material, por lo que obtenemos lo siguiente:  
- S.A
$$\frac{7}{10}x_{1} + x_{2} \leq 630$$
$$\frac{1}{2}x_{1} + \frac{5}{6}x_{2} \leq 600 $$
$$x_{1} + \frac{2}{3}x_{2} \leq 708 $$
$$\frac{1}{10}x_{1} + \frac{1}{4}x_{2} \leq 135 $$
> Para la cantidad maxima de producción de bolsas de en G&L.

In [102]:
# vectors of prove
# A is a matrix
A1 = np.array([[7/10, 1],
             [1/2, 5/6],
             [1, 2/3],
             [1/10,1/4]])

# b is the vector result
b1 = np.array([630, 600, 708, 135])

In [112]:
# Así pues se realiza el siguiente
# procedimiento de optimización 
# para el sistema y función de utilidad.
# P -> función de costo
# coeficientes de P 
c_ex = np.array([7,9]) # x_1 x_2 
# constantes de inecuaciones x_1 x_2 = 2 variables de control
# representan 4 restricciones
# [1] 
# [2] 
# [3] 
# [4]
A = np.array([[7/10, 1],
             [1/2, 5/6],
             [1, 2/3],
             [1/10,1/4]])
# se cre el vector solucion de la matriz
b = np.array([630, 600, 708, 135])
# restricciones de variables no negativas, (no puede existir producción negativa)
bounds_ex = [(  0,    None)]*2
# Solución de problema a traves de simplex
# se toma el vector C como negativo para encontrar el minimo de la funcion invertida, osea el máximo
res_ex1 = linprog(-c_ex, A_ub=A, b_ub=b, bounds=bounds_ex, method='revised simplex')
# se toma las soluciones probables del sistema 
res_ex1
# se toma los valores enteros menores para no estar por fuera la zona
# solucion

     con: array([], dtype=float64)
     fun: -6048.0
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([  0., 120.,   0.,  18.])
  status: 0
 success: True
       x: array([540., 252.])

In [118]:
# Así pues se realiza el siguiente
# procedimiento de optimización 
# para el sistema y función de utilidad.
# P -> función de costo
# coeficientes de P 
c_ex2 = np.array([7,0]) # x_1 x_2 
# constantes de inecuaciones x_1 x_2 = 2 variables de control
# representan 4 restricciones
# [1] 
# [2] 
# [3] 
# [4]
A = np.array([[7/10, 1],
             [1/2, 5/6],
             [1, 2/3],
             [1/10,1/4]])
# se cre el vector solucion de la matriz
b = np.array([630, 600, 708, 135])
# restricciones de variables no negativas, (no puede existir producción negativa)
bounds_ex = [(  0,    None)]*2
# Solución de problema a traves de simplex
# se toma el vector C como negativo para encontrar el minimo de la funcion invertida, osea el máximo
res_ex2 = linprog(-c_ex2, A_ub=A, b_ub=b, bounds=bounds_ex, method='revised simplex')
# se toma las soluciones probables del sistema 
res_ex2
# se toma los valores enteros menores para no estar por fuera la zona
# solucion

     con: array([], dtype=float64)
     fun: -4956.0
 message: 'Optimization terminated successfully.'
     nit: 1
   slack: array([134.4, 246. ,   0. ,  64.2])
  status: 0
 success: True
       x: array([708.,   0.])

In [119]:
# Así pues se realiza el siguiente
# procedimiento de optimización 
# para el sistema y función de utilidad.
# P -> función de costo
# coeficientes de P 
c_ex3 = np.array([0,9]) # x_1 x_2 
# constantes de inecuaciones x_1 x_2 = 2 variables de control
# representan 4 restricciones
# [1] 
# [2] 
# [3] 
# [4]
A = np.array([[7/10, 1],
             [1/2, 5/6],
             [1, 2/3],
             [1/10,1/4]])
# se cre el vector solucion de la matriz
b = np.array([630, 600, 708, 135])
# restricciones de variables no negativas, (no puede existir producción negativa)
bounds_ex = [(  0,    None)]*2
# Solución de problema a traves de simplex
# se toma el vector C como negativo para encontrar el minimo de la funcion invertida, osea el máximo
res_ex3 = linprog(-c_ex3, A_ub=A, b_ub=b, bounds=bounds_ex, method='revised simplex')
# se toma las soluciones probables del sistema 
# key = 'fun'
res_ex3
# se toma los valores enteros menores para no estar por fuera la zona
# solucion

     con: array([], dtype=float64)
     fun: -4860.0
 message: 'Optimization terminated successfully.'
     nit: 1
   slack: array([ 90., 150., 348.,   0.])
  status: 0
 success: True
       x: array([  0., 540.])

In [120]:
tabla = {'Tipo':['Normal','B.Estandar','B.Deluxe'], 'Valor':[], 'Solución':[]}
tabla['Valor'].append(res_ex1.get('x'))
tabla['Valor'].append(res_ex2.get('x'))
tabla['Valor'].append(res_ex3.get('x'))
tabla['Solución'].append(-1*res_ex1.get('fun'))
tabla['Solución'].append(-1*res_ex2.get('fun'))
tabla['Solución'].append(-1*res_ex3.get('fun'))

In [121]:
pd.set_option('max_colwidth',75)
df_sol = pd.DataFrame(tabla)
print('\n Soluciones de costos (Matriz de rango completo): \n')
display(df_sol)


 Soluciones de costos (Matriz de rango completo): 



Tipo                                   Valor  Solución
0      Normal  [539.9999999999999, 252.0000000000001]    6048.0
1  B.Estandar                            [708.0, 0.0]    4956.0
2    B.Deluxe                            [0.0, 540.0]    4860.0

> Para el Dual tenemos el siguiente planteamiento:
$$\max 630y_1 + 600y_2 + 708y_3 + 135y_4 = G(y_1, y_2, y_3, y_4)$$
> Sujeto a las siguientes restricciones
- S.A
$$\frac{7}{10}y_{1} + \frac{1}{2}y_{2} + y_3 + \frac{1}{10}y_{4}\leq -7$$
$$y_{1} + \frac{5}{6}y_{2} + \frac{2}{3}y_3 + \frac{1}{4}y_{4}\leq -9$$
> En forma estandar 
$$-\frac{7}{10}y_{1} + -\frac{1}{2}y_{2} + y_3 + -\frac{1}{10}y_{4}\geq 7$$
$$-y_{1} + -\frac{5}{6}y_{2} + -\frac{2}{3}y_3 + -\frac{1}{4}y_{4}\geq 9$$
> Donde lo que se busca para este problema $\textit{a}$ es aumentar la utilidad de la bolsa deluxe a \$ 18, por lo que modificaria las restricciones de la siguiente manera:
- S.A
$$\frac{7}{10}y_{1} + \frac{1}{2}y_{2} + y_3 + \frac{1}{10}y_{4}\leq 7$$
$$y_{1} + \frac{5}{6}y_{2} + \frac{2}{3}y_3 + \frac{1}{4}y_{4}\leq 18$$
> Donde lo que se busca para este problema $\textit{b}$ es aumentar la utilidad de la bolsa deluxe a \$ 18, por lo que modificaria las restricciones de la siguiente manera:
- S.A
$$\frac{7}{10}y_{1} + \frac{1}{2}y_{2} + y_3 + \frac{1}{10}y_{4}\leq 14$$
$$y_{1} + \frac{5}{6}y_{2} + \frac{2}{3}y_3 + \frac{1}{4}y_{4}\leq 9$$
> Donde lo que se busca para este problema $\textit{c}$ es aumentar la utilidad de la bolsa deluxe a \$ 18, por lo que modificaria la función de costo de la siguiente manera:
$$1.7*[630y_1 + 600y_2 + 708y_3 + 135y_4] = P(y_1, y_2, y_3, y_4)$$


In [186]:
# Solución de problema DUAL
res_ex4 = linprog(b, np.transpose(-A), -c_ex, method='revised simplex')
# se toma las soluciones probables del sistema 
# key = 'fun'
res_ex4
# se toma los valores enteros menores para no estar por fuera la zona
# solucion

     con: array([], dtype=float64)
     fun: 6048.0
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([0., 0.])
  status: 0
 success: True
       x: array([8.125 , 0.    , 1.3125, 0.    ])

> En este caso nuevo caso se busca aumentar la contribución total de G&L a $18, mientras que el modo estándar si continua con una utilidad de $7.

In [195]:
# Solución de problema DUAL
c_new1 = np.array([7,18])
res_ex5 = linprog(b, np.transpose(-A), -c_new1, method='revised simplex')
# se toma las soluciones probables del sistema 
# key = 'fun'
res_ex5
# se toma los valores enteros menores para no estar por fuera la zona
# solucion

     con: array([], dtype=float64)
     fun: 9720.0
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([0.2, 0. ])
  status: 0
 success: True
       x: array([ 0.,  0.,  0., 72.])

> También, se empleará una materia prima de e bajo costo para la fabricación del modelo estándar para que se pueda incrementar su utilidad por bolsa a $14, para este caso la utilidad por bolsa del modelo Deluxe permanecerá en $9. 

In [196]:
# Solución de problema DUAL
c_new2 = np.array([14,9])
res_ex6 = linprog(b, np.transpose(-A), -c_new2, method='revised simplex')
# se toma las soluciones probables del sistema 
# key = 'fun'
res_ex6
# se toma los valores enteros menores para no estar por fuera la zona
# solucion

     con: array([], dtype=float64)
     fun: 9911.999999999998
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([-1.77635684e-15,  3.33333333e-01])
  status: 0
 success: True
       x: array([ 0.,  0., 14.,  0.])

> Finalmente se adquiere más personal en el equipo de trabajo para aumentar las horas de trabajo disponible en un 70%.

In [204]:
# Solución de problema DUAL
res_ex7 = linprog(1.7*b, np.transpose(-A), -c_ex, method='revised simplex')
# se toma las soluciones probables del sistema 
# key = 'fun'
res_ex7
# se toma los valores enteros menores para no estar por fuera la zona
# solucion

     con: array([], dtype=float64)
     fun: 10281.6
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([0., 0.])
  status: 0
 success: True
       x: array([8.125 , 0.    , 1.3125, 0.    ])

> Al problema primal que se muestra a continuación se le busca implementar por medio de Python una solución a dicho problema primal y dual; de forma que, por medio de las ecuaciones de holgura complementaria, se puedan encontrar la cantidad de bolsas a fabricar tanto de modo estándar como como Deluxe.

In [286]:
def sol_hol_comp(A, b , f, res, dict, costos = np.array([7,9])):
    flag = False
    pos = []
    res_x = np.array([res])
    # B = np.matmul(np.transpose(f),A)
    c = np.matmul(np.transpose(f),b)
    if c == res:
        flag = True
    dict['Cumple'].append(flag)
    for i in range(len(f)):
        if f[i] == 0:
            pos.append(i)
    dict['Restricción(Inactivas)'].append(pos)
    if len(pos) <= len(A[0,:]):
        An = np.delete(A,(pos), axis = 0)
        bn = np.delete(b,(pos))
        sol = solve(An,bn)
    else:
        An = np.delete(A,(pos), axis = 0)
        An = np.vstack([An, costos])
        bn = np.delete(b,(pos))
        bn = np.concatenate((bn,res_x))
        sol = solve(An,bn)
    dict['Solución'].append(sol)
    return dict

> Se crea un diccionario para graficar los datos

In [287]:
if __name__ == "__main__":
    dict = {'Cumple':[],'Restricción(Inactivas)':[],'Solución':[]}
    h1 = sol_hol_comp(A1, b1, res_ex4.get('x'), res_ex4.get('fun'), dict)
    h2 = sol_hol_comp(A1, b1, res_ex5.get('x'), res_ex5.get('fun'), h1)
    h3 = sol_hol_comp(A1, b1, res_ex6.get('x'), res_ex6.get('fun'), h2)
    h4 = sol_hol_comp(A1, b1*1.7, res_ex7.get('x'), res_ex7.get('fun'), h3)

In [310]:
pd.set_option('max_colwidth',75)
df_sol = pd.DataFrame(h4)
print('\n Soluciones según el Dual: \n')
display(df_sol)


 Soluciones según el Dual: 



Cumple Restricción(Inactivas)                  Solución                
0   True            [1, 3]                     [540.0, 252.00000000000003]
1   True         [0, 1, 2]        [1429.4117647058824, -31.76470588235297]
2   True         [0, 1, 3]        [-54.46153846153824, 1143.6923076923074]
3   True            [1, 3]         [917.9999999999998, 428.40000000000015]

> Como podemos ver en la tabla anterior la única solución factible es la ultima, ya que, tiene soluciones basicas no negativas, maximizando la utilidad de la función de costo de la siguiente forma:
$$(|1-\frac{10281.6}{6048.0}|)*100\%=70\%$$
> Tuvo un aumento del 70\% en la función de costo.